[tsulim] iP

tsulim opened the pull request:
# [Seiki](#Seiki)
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Seiki frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/tsulim/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method in `Seiki.java`:
```
public static void main(String[] args) {
        new Seiki("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tsulim committed:
Level-0

tsulim committed:
Level-1

tsulim committed:
Level-2

tsulim committed:
Added new classes

tsulim committed:
Level-3 completed

tsulim committed:
Added new classes inheriting from Task

tsulim committed:
Level-4 completed

tsulim committed:
Added test input & expected output

tsulim committed:
A-TextUiTesting completed

tsulim committed:
Level-5 completed

tsulim committed:
Modified test input and expected output

tsulim committed:
Level-6 completed

tsulim committed:
Split codes to separate functions

tsulim committed:
Added enums

tsulim committed:
Add Save Function

tsulim committed:
Merge branch 'branch-Level-7'

tsulim committed:
Add Date and Time parsing

tsulim committed:
Modify test input and expected output

tsulim committed:
Merge branch 'branch-Level-8'

tsulim committed:
Stop tracking storage file

tsulim committed:
Make code more OOP and organize into packages

tsulim committed:
Merge branch 'branch-A-MoreOOP'

tsulim committed:
Fix bash file

tsulim committed:
Merge branch 'branch-A-Packages'

tsulim committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

tsulim committed:
Add Gradle Support

tsulim committed:
Refactor code to comply with the Checkstyle rules

tsulim committed:
Merge branch 'branch-A-Gradle'

tsulim committed:
Add try catch to parse method in Parser

tsulim committed:
Add JUnit tests

tsulim committed:
Merge branch 'branch-A-JUnit'

tsulim committed:
Refactor code to comply with the Checkstyle rules

tsulim committed:
Enable assertions in build.gradle

tsulim committed:
Merge branch 'branch-A-Jar'

tsulim committed:
Add Find command

tsulim committed:
Add JavaDocs to comply with the Checkstyle rules

tsulim committed:
Merge branch 'branch-A-JavaDoc'

tsulim committed:
Remove redundant comment

tsulim committed:
Merge branch 'branch-A-CodingStandard'

tsulim committed:
Merge branch 'branch-Level-9'

tsulim committed:
Add JavaDocs to comply with the Checkstyle rules

tsulim committed:
Merge branch 'branch-A-Checkstyle'

AL-ZT commented on src/main/java/seiki/data/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+     * @param tasks
+     */
+    public TaskList(ArrayList<Task> tasks) {
+        this.taskList = tasks;
+        this.taskCount = tasks.size();
```
Perhap 'this' keyword should be omitted?

tsulim replied:
Oh! It slipped my mind. Thank you for spotting it!

AL-ZT commented on src/test/java/data/task/EventTest.java:
```diff
@@ -0,0 +1,47 @@
+
+import seiki.data.task.Event;
+
+public class EventTest {
+    @Test
+    public void toString_unmarkedEvent_success() {
+        Event task = new Event("test",
+                LocalDateTime.parse("2020/02/02 1234", DATE_TIME_FORMATTER),
+                LocalDateTime.parse("2020/02/02 2234", DATE_TIME_FORMATTER));
+        assertEquals("[E] test (from: 02 Feb 2020 1234 to: 02 Feb 2020 2234)", task.toString());
+    }
+
+
+    @Test
+    public void toString_markedEvent_success() {
+        Event task = new Event("test",
+                LocalDateTime.parse("2020/02/02 1234", DATE_TIME_FORMATTER),
+                LocalDateTime.parse("2020/02/02 2234", DATE_TIME_FORMATTER));
+        task.markAsDone();
+        assertEquals("[E] test (from: 02 Feb 2020 1234 to: 02 Feb 2020 2234) ✓", task.toString());
+    }
+
+    @Test
+    public void toFile_unmarkedEvent_success() {
+        Event task = new Event("test",
+                LocalDateTime.parse("2020/02/02 1234", DATE_TIME_FORMATTER),
+                LocalDateTime.parse("2020/02/02 2234", DATE_TIME_FORMATTER));
+        assertEquals("E | 0 | test | 2020/02/02 1234-2020/02/02 2234", task.toFile());
+    }
+
+    @Test
+    public void toFile_markedEvent_success() {
+        Event task = new Event("test",
+                LocalDateTime.parse("2020/02/02 1234", DATE_TIME_FORMATTER),
+                LocalDateTime.parse("2020/02/02 2234", DATE_TIME_FORMATTER));
+        task.markAsDone();
+        assertEquals("E | 1 | test | 2020/02/02 1234-2020/02/02 2234", task.toFile());
```
Good job adding more unit tests than what was required!

AL-ZT commented on src/main/java/seiki/commands/EventCommand.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    public void execute(Storage storage, TaskList taskList, Ui ui) throws SeikiException {
+        if (!this.args.contains("/from") && !this.args.contains("/to")) {
+            throw new SeikiException("The date time for the task is missing.\n"
+                    + "Please use the following format: event [task title] /from [startdatetime] /to [enddatetime]");
+        } else if (this.args.contains("/from")
+                && this.args.contains("/to")
+                && (this.args.subList(0, this.args.indexOf("/from")).size() == 0
+                && this.args.subList(this.args.indexOf("/to") + 1, this.args.size()).size() == 0
+                && this.args.subList(this.args.indexOf("/from") + 1, this.args.indexOf("/to")).size() == 0)) {
+            throw new SeikiException("The task title, start and end date time for the task is missing.\n"
+                    + "Please use the following format: event [task title] /from [startdatetime] /to [enddatetime]");
+        } else if (!this.args.contains("/to")
+                || this.args.subList(this.args.indexOf("/to") + 1, this.args.size()).size() == 0) {
+            throw new SeikiException("The date time for the task is missing.\n"
+                    + "Please use the following format: event [task title] /from [startdatetime] /to [enddatetime]");
+        } else if (!this.args.contains("/from")
+                || this.args.subList(this.args.indexOf("/from") + 1, this.args.indexOf("/to")).size() == 0) {
+            throw new SeikiException("The date time for the task is missing.\n"
+                    + "Please use the following format: event [task title] /from [startdatetime] /to [enddatetime]");
+        } else if (this.args.subList(0, this.args.indexOf("/from")).size() == 0) {
+            throw new SeikiException("The task title is missing.\n"
+                    + "Please use the following format: event [task title] /from [startdatetime] /to [enddatetime]");
+        } else {
+            String taskName = String.join(" ", this.args.subList(0, this.args.indexOf("/from")));
+            String startDateTimeStr = String.join(" ",
+                    this.args.subList(this.args.indexOf("/from") + 1, this.args.indexOf("/to")));
+            String endDateTimeStr = String.join(" ",
+                    this.args.subList(this.args.indexOf("/to") + 1, this.args.size()));
+            try {
+                LocalDateTime startDateTime = LocalDateTime.parse(startDateTimeStr, DATE_TIME_FORMATTER);
+                LocalDateTime endDateTime = LocalDateTime.parse(endDateTimeStr, DATE_TIME_FORMATTER);
+                Event task = new Event(taskName, startDateTime, endDateTime);
+                taskList.addTask(task);
+                storage.save(taskList);
+                ui.showAddTask(task, taskList);
+            } catch (DateTimeParseException e) {
+                throw new SeikiException("The format of the date time is incorrect.\n"
+                        + "Please use the following format: yyyy/MM/dd HHmm");
```
Could the execution of Event be further improved to remove clutter?

eliotlee68 commented on src/main/java/seiki/data/task/ToDo.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Represents a todo task.
+ */
+public class ToDo extends Task {
```
Maybe TodoTask (and similarly for event and deadline) would be class naming that implies the task inheritance

tsulim replied:
I understand where you are coming from. Thanks!

AL-ZT commented on src/main/java/seiki/parser/Parser.java:
```diff
@@ -0,0 +1,102 @@
+                    throw new SeikiException("The task title is missing.\n"
+                            + "Please use the following format: todo [task title]");
+                }
+                return new ToDoCommand(args);
+            case DeadlineCommand.COMMAND_WORD:
+                if (args.isEmpty() || (args.subList(args.indexOf("/by") + 1, args.size()).isEmpty()
+                        && args.subList(0, args.indexOf("/by")).isEmpty())) {
+                    throw new SeikiException("The task title and date time for the task is missing.\n"
+                            + "Please use the following format: deadline [task title] /by [datetime]");
+                }
+                return new DeadlineCommand(args);
+            case EventCommand.COMMAND_WORD:
+                if (args.isEmpty()) {
+                    throw new SeikiException("The task title, start and end date time for the task is missing.\n"
+                            + "Please use the following format:"
+                            + "event [task title] /from [startdatetime] /to [enddatetime]");
+                }
+                return new EventCommand(args);
+            case DeleteCommand.COMMAND_WORD:
+                if (args.isEmpty()) {
+                    throw new SeikiException("Please enter a task number.");
+                } else if (args.size() != 1) {
+                    throw new SeikiException("Multiple inputs have been detected.\n"
+                            + "Please follow the format: delete [task number]");
+                }
+                return new DeleteCommand(args);
+            case ByeCommand.COMMAND_WORD:
+                return new ByeCommand();
+            case FindCommand.COMMAND_WORD:
+                if (args.isEmpty()) {
+                    throw new SeikiException("Please enter a keyword.");
+                } else if (args.size() != 1) {
+                    throw new SeikiException("Please enter only 1 keyword.");
+                }
+                return new FindCommand(args);
+            default:
+                throw new SeikiException("I'm sorry, I didn't quite understand that.");
+            }
+        } catch (IndexOutOfBoundsException e) {
+            throw new SeikiException("I'm sorry, I didn't quite understand that.");
```
Maybe more extensions of your SeikiException would be ideal in handling different cases of inputs?

eliotlee68 commented on src/main/java/seiki/ui/Ui.java:
```diff
@@ -0,0 +1,144 @@
+        }
+    }
+
+    public String readCommand() {
+        return in.nextLine();
+    }
+
+    public void showLine() {
+        showToUser(DIVIDER);
+    }
+
+    public void showError(String message) {
+        showToUser(message);
+    }
```
Javadocs would be appreciated for these methods

tsulim replied:
Okay, I will add them later on.

eliotlee68 reviewed (commented):
LGTM, just minor nitpicks about javadocs and class naming conventions

AL-ZT commented on src/main/java/seiki/common/Messages.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public class Messages {
+
+    public static final String MESSAGE_GREETING = "Hello! I'm Seiki\nHow may I assist you today?";
+    public static final String MESSAGE_FAREWELL = "Goodbye! If you ever need assistance in the future, "
+                                                + "don't hesitate to reach out. Take care!";
+    public static final String MESSAGE_LOGO = "  _____      _  _     _\n"
+                                            + " / ____|    (_)| | _ (_)\n"
+                                            + "| (___  ___  _ | |/ / _\n"
+                                            + " \\___ \\/ _ \\| ||   / | |\n"
+                                            + " ____) | __/| || | \\ | |\n"
+                                            + "|_____/\\___||_||_|\\_\\|_|\n";
+    public static final String MESSAGE_ADD_SUCCESS = "The following task has been added:";
+    public static final String MESSAGE_DELETE_SUCCESS = "The following task has been deleted.";
+    public static final String MESSAGE_FIND_SUCCESS = "These are the following tasks that matches the keyword: %1$s";
+    public static final String MESSAGE_LIST_SUCCESS = "Here are the task(s) you have:";
+    public static final String MESSAGE_MARK_SUCCESS = "The following task has been marked.";
+    public static final String MESSAGE_REMAINING_TASKS = "You have a total of %1$d tasks in the list.";
+    public static final String MESSAGE_TASK = "→ %1$s";
+    public static final String MESSAGE_UNMARK_SUCCESS = "The following task has been unmarked.";
```
I like your effort with maintaining reused messages throughout your application!

AL-ZT reviewed (commented):
Overall, your code was mostly following the coding standard and I really like your different package layouts. I would suggest try to find better ways to parse and execute your commands when you have the time, but good formatting!

tsulim committed:
Add GUI

tsulim committed:
Merge branch 'branch-Level-10'

tsulim committed:
Modify build.gradle

tsulim committed:
Rename ToDo, Deadline and Event classes with Task suffix

The class names ToDo, Deadline and Event do not accurately convey their
inheritance as tasks.

Using names like ToDoTask, DeadlineTask and EventTask provides clearer
indication of their role as task objects.

Rename the ToDo class to ToDoTask, Deadline class to DeadlineTask, and
Event class to EventTask. Update all references and dependencies
accordingly.

By appending "Task" to the class names, it explicitly communicates their
inheritance as tasks, enhancing code readability and maintainability.

Ensure thorough testing after the changes to confirm the continued
functionality of the codebase.

tsulim committed:
Add JavaDocs to the remaining method in Ui class for consistency

Some methods in the class lack JavaDocs comments, causing inconsistency
in documentation.

Consistent documentation improves code readability, maintainability,
and understanding. Having JavaDocs comments for all methods ensures
that developers have complete information about each method's purpose,
parameters, and behaviour.

Add JavaDocs comments to the remaining method in Ui class to
maintain consistency in documentation.

By adding JavaDocs comments to all methods, we ensure that the
documentation is comprehensive and consistent throughout the class.
This approach facilitates easier understanding and usage of the class
by developers.

tsulim committed:
Remove unused variable from Ui class

The Ui class includes a declaration for the private static final String
DIVIDER, which is set to a string value.

The DIVIDER string is not being used within the Ui class and appears
to be redundant. Removing it will declutter the code and improve its
readability.

Remove the declaration for the private static final String DIVIDER from
the Ui class.

Removing unused variables helps to simplify the codebase and reduces
potential confusion for developers who may wonder about the purpose
of the variable.

tsulim committed:
Omit the 'this' keyword from class files where redundant

Some class files contain instances of the 'this' keyword that are
redundant and unnecessary.

The 'this' keyword is used in Java to refer to the current object
instance. However, in certain contexts, its usage is redundant and adds
unnecessary verbosity to the code.

Omit the 'this' keyword from class files where redundant.

Removing redundant 'this' keywords simplifies the code and improves
readability by eliminating unnecessary noise.

Ensure that the removal of 'this' keywords does not alter the behaviour
or functionality of the classes. Review the affected code to confirm
that it remains clear and understandable after the changes.

tsulim committed:
Add Assertions

tsulim committed:
Merge pull request #2 from tsulim/branch-A-Assertions

Add Assertions to code

tsulim committed:
Refactor code to improve Code Quality

tsulim committed:
Merge master into branch-A-CodeQuality

tsulim committed:
Merge pull request #3 from tsulim/branch-A-CodeQuality

Refactor code to improve Code Quality

tsulim committed:
Add Help Function

tsulim committed:
Merge branch 'branch-BCD-Extension'

tsulim committed:
Finalize code

tsulim committed:
Add a user guide to the project
